File: DataDriven/expressJS/src/controllers/Comment.js
const Comment = require("../models/Comment");
const Post = require("../models/Post");

exports.createBulk = async (req, res) => {
  const commentsData = Object.assign([], req.body.commentsData);
  if (!Array.isArray(commentsData.commentsData) && commentsData.length < 1) {
    return res.status(422).json({ error: "Invalid request data." });
  }
  try {
    const commentSavingAsync = commentsData.map(async (commentInput) => {
      const post = await Post.findOne({
        id: commentInput.post,
      }).exec();
      if (!post) throw new Error("Post not found! " + commentInput.post);
      commentInput.post = post._id;
      commentInput.id = 5;
      const savedComment = await Comment.create(commentInput);
      if (!savedComment) {
        throw new Error("Error creating database entry!");
      }
    });
    Promise.all(commentSavingAsync)
      .then(() => {
        res.status(200).json({ success: true });
      })
      .catch((e) => {
        res.status(500).json({ error: `${e}` });
      });
  } catch (e) {
    res.status(500).json({ error: `${e}` });
  }
};

exports.deleteBulk = async (req, res) => {
  const { postIDs } = req.body;
  if (!Array.isArray(postIDs) || postIDs.length < 1) {
    return res.status(422).json({ error: "Invalid request data." });
  }
  try {
    const posts = await Post.find({ id: { $in: postIDs } }).exec();
    if (posts.length < 1) {
      return res.status(200).json({ result: "No ppsts found." });
    }
    const postMongoIDs = posts.map(({ _id }) => _id);
    const deleted = await Comment.deleteMany({ post: { $in: postMongoIDs } });
    if (Object.hasOwn(deleted, "acknowledged") && deleted.acknowledged) {
      res.status(200).json({ success: true });
    } else {
      throw new Error("Error deleting database entry!");
    }
  } catch (e) {
    res.status(500).json({ error: `${e}` });
  }
};

File: DataDriven/Gin/internal/server/controllers/comment.go
package controllers

import (
	"net/http"
	"time"

	"Gin/internal/models"

	"github.com/gin-gonic/gin"
)

type commentInput struct {
	AuthorName  string `json:"authorName"`  // Author name, max length 50, required
	Content     string `json:"content"`     // Content of the comment, required
	LastUpdated string `json:"lastUpdated"` // Last updated time, defaults to current timestamp
	Post        uint   `json:"post"`        // Related Post model
}

func (s *Server) CreateBulkCommentsHandler(c *gin.Context) {
	var req struct {
		CommentsData []commentInput `json:"commentsData"`
	}

	if err := c.BindJSON(&req); err != nil {
		c.JSON(http.StatusUnprocessableEntity, gin.H{"error": "Invalid request data."})
		return
	}

	if len(req.CommentsData) < 1 {
		c.JSON(http.StatusUnprocessableEntity, gin.H{"error": "Invalid request data."})
		return
	}

	var comments []models.Comment

	for _, commentData := range req.CommentsData {
		lastUpdated, err := time.Parse("2006-01-02", commentData.LastUpdated)
		if err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "invalid date format for LastUpdated"})
			return
		}

		comment := models.Comment{
			AuthorName:  commentData.AuthorName,
			Content:     commentData.Content,
			LastUpdated: lastUpdated,
			PostID:      commentData.Post,
		}

		comments = append(comments, comment)
	}

	if err := s.db.Create(&comments).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"success": true})
}

func (s *Server) DeleteBulkCommentsHandler(c *gin.Context) {
	var req struct {
		PostIDs []uint `json:"postIDs"`
	}

	if err := c.BindJSON(&req); err != nil {
		c.JSON(http.StatusUnprocessableEntity, gin.H{"error": "Invalid request data."})
		return
	}

	if len(req.PostIDs) < 1 {
		c.JSON(http.StatusUnprocessableEntity, gin.H{"error": "Invalid request data."})
		return
	}

	var posts []models.Post
	if err := s.db.Where("id IN (?)", req.PostIDs).Find(&posts).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	if len(posts) < 1 {
		c.JSON(http.StatusOK, gin.H{"result": "No posts found."})
		return
	}

	var postIDs []uint
	for _, post := range posts {
		postIDs = append(postIDs, post.ID)
	}

	if err := s.db.Unscoped().Where("post_id IN (?)", postIDs).Delete(&models.Comment{}).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"success": true})
}
